[Pytorch][Annotation] Use call stack to get module path

cccclai opened the pull request:
### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.
5. Update the unit test `test/mobile/test_lite_script_module.py`, and add the test to cover module produced by torch.jit.script(). With this pr, the module level information can be generated for the scripted models as well.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46893 [Pytorch][Annotation] Use call stack to get module path**
* #47420 [Pytorch][Annotation] Update inlined callstack with module instance info

Differential Revision: [D24752847](https://our.internmc.facebook.com/intern/diff/D24752847)

cccclai committed:
[Pytorch][Annotation] Use call stack to get module path

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"




[ghstack-poisoned]

iseeyuan commented:
There are two submodules in this PR. Did you add them by accident?

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -54,20 +54,72 @@ static IValue Table(
+  return callStack.has_value() ? callStack->get()->getModulePath() : "";
+}
+
+std::string getModuleTypeName(const Module& module, const std::string& prefix) {
```
nit: If it's only used in this file, please use anonymous namespace to mark it "local".

iseeyuan replied:
nvm. It is already in anonymous namespace. My bad...

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -54,20 +54,72 @@ static IValue Table(
+  return prefix + "(" + moduleType + ")";
+}
+
+void setModulePath(Node* node, const std::string& root_scope_string) {
```
Same here. Put it inside namespace {}

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -54,20 +54,72 @@ static IValue Table(
+  }
+}
+
+void setModulePaths(
```
In anonymous namespace.

iseeyuan reviewed (approved):
LGTM. Please rebase to resolve the CI failures.

iseeyuan commented:
@kimishpatel , could you take a look to make sure it's expected?

dhruvbird commented:
@cccclai  I noticed that your changes also touch some submodules such as `third_party/XNNPACK`, etc... I believe if you have these changes you won't be able to push your commit (@iseeyuan thoughts?)

dhruvbird commented on torch/csrc/jit/ir/scope.cpp:
```diff
@@ -123,12 +125,23 @@ std::vector<InlinedCallStackEntry> InlinedCallStack::vec() {
 }
 
+void InlinedCallStack::setModulePath(std::string path) {
+  modulePath_ = path;
```
modulePath_ = std::move(path);

dhruvbird commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -112,8 +111,12 @@ struct TORCH_API InlinedCallStack : public c10::intrusive_ptr_target {
   InlinedCallStackPtr intrusive_from_this();
   c10::optional<ModuleInstanceInfo> module_instance_info_;
+  std::string modulePath_;
+  bool initialized_ = true;
```
Q. Where are you using the `initialized_` variable?

cccclai replied:
`InlinedCallStack ` is an optional field. If a node did not come from any inlined function, its callstack will be empty and modulePath_ can't be set.I was thinking to use initialized field to indicate its validity, instead of using the x.has_value() from std::optional. Anyway, I will take @kimishpatel suggestion and use std::unordered_map, such that we don't need to add extra field in node.

iseeyuan commented:
> @cccclai I noticed that your changes also touch some submodules such as `third_party/XNNPACK`, etc... I believe if you have these changes you won't be able to push your commit (@iseeyuan thoughts?)

Indeed. The submodules need to be removed before committing it.

kimishpatel commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1928,7 +1928,10 @@ std::vector<Value*> inlineCallTo(
+                *new_node_cs,
+                callee,
+                to_replace->sourceRange(),
+                module_instance_info);
```
If possible formatting changes should be kept separate.

kimishpatel commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -326,6 +326,11 @@ struct TORCH_API Node {
+  void setModulePathStr(std::string modulePathStr) {
+    callstack_->get()->setModulePath(modulePathStr);
+  }
+
```
I am vary of introducing extra methods to Node class if it can be avoided. I dont think we need it for this. See my other comments.

kimishpatel commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -140,6 +143,8 @@ struct TORCH_API InlinedCallStack : public c10::intrusive_ptr_target {
 
   // Return callstack as a vector of [Function, SourceRange] pairs.
   std::vector<InlinedCallStackEntry> vec();
+  void setModulePath(std::string path);
+  std::string getModulePath();
```
I think we can avoid this all together. Especially in the context that inlinedcallstack knows only about the function it is in and where in that function. In that we added what module that function is in and what instance this particular instance of inlinedcallstack corresponds to. Thus what is all the entry in the callstack put together look like should be outside of this class. So I don't think we need `setModulePath` and `getModulePath` functions. See my later comments as I am typing them.

kimishpatel commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1633,8 +1633,8 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
         for (const auto& p : (*node->callstack())->vec()) {
-          entries.emplace_back(StackEntry{previous_fn_name, p.second});
-          previous_fn_name = p.first->name();
+          entries.emplace_back(StackEntry{previous_fn_name, std::get<1>(p)});
+          previous_fn_name = std::get<0>(p)->name();
```
@ZolotukhinM, are you ok with these changes? It seems from last time you commented on the fact that InlinedCallStack's `vec` method has no other users, to now it has changed. And it seems it is being used here. Do you prefer this change or add another API to InlinedCallStack.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -54,20 +54,72 @@ static IValue Table(
+      }
+    }
+
+    node->setModulePathStr(module_info);
```
So what I am thinking is that instead of doing this we can just have `std::unordered_map<Node*, std::string> node_module_paths;` and basically we just do `node_module_paths.emplace(node, module_info)`. Then you `getModulePath` will be `return node_module_paths[node]` in which case you may not need that function at all. You can have node_module_path created in getFunctionTuple and passed down from setModulePaths.

cccclai replied:
That's a good idea. I tried it and it seems not working out of box. After looking into it, the nodes generated from `code.instructions_source()[i]` are from `PreprocessGraph`, which is a graph copy and the node pointer will be different. If it's expected, what will be the ideal key to find the unique node?

kimishpatel replied:
I see. Yes that is bit troubling. One thing I can think of is moving these lines, https://github.com/pytorch/pytorch/blob/5dad1d541229f97d386967eb177baa5c1740e374/torch/csrc/jit/serialization/export_module.cpp#L132-L135, here: https://github.com/pytorch/pytorch/blob/5dad1d541229f97d386967eb177baa5c1740e374/torch/csrc/jit/serialization/export_module.cpp#L147. So instead of doing 
```
  if (save_mobile_debug_info) {
    std::string root_scope_string = getModuleTypeName(module, "top");
    setModulePaths(graph, root_scope_string);
  }
```
you actually do
```
  if (save_mobile_debug_info) {
    std::string root_scope_string = getModuleTypeName(module, "top");
    setModulePath(node, root_scope_string);
  }
```
I think some variant of this should work, since we are not really using graph to populate debug info. Instead we are using `Node`'s callstack to populate debug info. So if you can just use node to build module_debug_info that is even better. We do not have to maintain any map, we get the debug info right when we need it.

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -505,7 +505,7 @@ TEST(LiteInterpreterTest, SequentialModuleInfo) {
 
   std::unordered_set<std::string> expected_result(
-      {"top(C).A0(A).forward", "top(C).B0(B).forward"});
+      {"top(C).A0(A).forward", "top(C).A0(A).B0(B).forward"});
```
Why this test case was changed? Give B0(B) and A0(A) are inside C you should not see top(C).A0(A).B0(B).

cccclai replied:
Module C forward methods is return self.A0.forward(self.B0.forward(x)), so the expected result should be top(C).A0(A).B0(B).forward, do I understand correctly?

kimishpatel commented on test/mobile/test_lite_script_module.py:
```diff
@@ -63,6 +63,33 @@ def forward(self, x):
+        assert(b"top(B).forward" in bytes)
+        assert(b"top(B).A0(A).forward" in bytes)
+        assert(b"top(B).A1(A).forward" in bytes)
+
```
Can you also add testcase for:
```
import torch
import torch.nn as nn

class A(nn.Module):
  def __init__(self):
    super(A, self).__init__()

  def forward(self, x):
    return x + 1

class B(nn.Module):
  def __init__(self):
    super(B, self).__init__()

  def forward(self, x):
    return x + 2

class C(nn.Module):
  def __init__(self):
    super(C, self).__init__()
    self.A0 = A()
    self.B0 = B()

  def forward(self, x):
    return self.A0(self.B0(x)) + 1
```
You should still expect to see (top(C).B0(B).forward) and (top(C).A0(A).forward)

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -54,20 +54,72 @@ static IValue Table(
+          auto type_name = class_type->name()->qualifiedName();
+          type_name = type_name.substr(type_name.find_last_of(".") + 1);
+          module_info += "." + instance_name + "(" + type_name + ")";
+          if (i == vec.size() - 1) {
```
I think we can append method name for all the nodes and not just the last one. Because otherwise we are making assumption that everything else inbetween calls forward method only. If this was in the base PR it probably was an error made either by me or Tai.
Then you will likely see things like "top(C).forward.B0(B).forward" etc.
Also can yo also import test cases from the base PR: 42174, as there are more cases being covered tehre? If you feel differently then please comment.

cccclai replied:
Address the change in https://github.com/pytorch/pytorch/pull/47416. Also update the extected_result in unittest.

kimishpatel reviewed (changes requested):
So I left a few comments. Overall there are few changes I want to make that minimizes any code changes you need to make to core JIT. Also things that can be avoided without adding new APIs are always good.

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.
5. Update the unit test `test/mobile/test_lite_script_module.py`, and add the test to cover module produced by torch.jit.script(). With this pr, the module level information can be generated for the scripted models as well.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.
5. Update the unit test `test/mobile/test_lite_script_module.py`, and add the test to cover module produced by torch.jit.script(). With this pr, the module level information can be generated for the scripted models as well.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.
5. Update the unit test `test/mobile/test_lite_script_module.py`, and add the test to cover module produced by torch.jit.script(). With this pr, the module level information can be generated for the scripted models as well.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.
5. Update the unit test `test/mobile/test_lite_script_module.py`, and add the test to cover module produced by torch.jit.script(). With this pr, the module level information can be generated for the scripted models as well.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Use call stack to get module path"

### Summary

This pr covers a couple of changes and remove the dependency on scope:
1. Update `InlinedCallStackEntry` from `std::pair<Function*, SourceRange>` to `std::tuple<Function*, SourceRange, c10::optional<ModuleInstanceInfo>`, per suggestion in the pr #42174. Update the syntax accordingly.
2. Update `reconstructScopes` (update scope for each node) function to `setModulePaths` (update the module path in callstack for each node).
3. Update `getModulePath` to use the `node->callStack->get()->getModulePath()`, instead of ` node->scopeName()`.
4. Add a field `initialized_` and `modulePath_` in `InlinedCallStack`. Otherwise, If a node did not come from any inlined function, its callstack will be empty and `modulePath_` can't be set.
5. Update the unit test `test/mobile/test_lite_script_module.py`, and add the test to cover module produced by torch.jit.script(). With this pr, the module level information can be generated for the scripted models as well.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=* --gtest_color=no

### Next Step
1. Remove the reconstruct_scope.cpp/reconstruct_scope.h files, since they are not needed any more.
2. In `ir.h`, use InlinedCallStackPtr callstack_ instead of `c10::optional<InlinedCallStackPtr> callstack_`. Use `initialized_` when checking value validity.
3. Unit test.




[ghstack-poisoned]

cccclai commented:
This pr seems stacked on an old commit. I folded the change and the new commit generates an new pr. (https://github.com/pytorch/pytorch/pull/47416). The new pr covers the requested changes.
